### PR TITLE
feat: adding pool retirement operations

### PR DESCRIPTION
**Pull request by abarretoatix**
# Description

Adds support for pool retirements at **construction api**.

# Proposed Solution

A new type of operation, **pool retirement** was added to support these kind of operation in a transaction.

# Testing

New tests were introduced in order to cover new test cases introduced.

# References

Closes construction api issue descripted at #342

### CODE REVIEW

File: cardano-rosetta-server/src/server/utils/cardano/operations-processor.ts [hunk 1]
```diff
@@ -182,6 +183,25 @@ const processOperationCertification = (
...
+    const poolKeyHash: string = operation.account?.address;
+    const epoch: number = operation.metadata.epoch;
+    const keyHash = CardanoWasm.Ed25519KeyHash.from_bytes(Buffer.from(poolKeyHash, 'hex'));
+    return {
+      certificate: CardanoWasm.Certificate.new_pool_retirement(PoolRetirement.new(keyHash, epoch)),
+      poolKeyHash
+    };
+  }
+  logger.error('[processPoolRetiring] Operation metadata information is missing');
+  throw ErrorFactory.missingMetadataParametersForPoolRetirement();
```
Comment (jvieiro) [inline cardano-rosetta-server/src/server/utils/cardano/operations-processor.ts:202]: I think this error only applies when the epoch parameter is missing. To validate the address you can use the existing function called `validateAndParsePoolKeyHash` that also parses the value.

**Review by jvieiro — CHANGES_REQUESTED** [review]
retirement operations should be added to output's related operations at `mapToRosettaTransaction` in `data-mapper`

File: cardano-rosetta-server/src/server/utils/cardano/transactions-processor.ts [hunk 2]
```diff
@@ -132,10 +132,31 @@ const parseCertToOperation = (
       staking_credential: { hex_bytes: hash, curve_type: CurveType.edwards25519 }
     }
   };
-  const delegationCert = cert.as_stake_delegation();
-  if (delegationCert) {
+  if (type === OperationType.STAKE_DELEGATION) {
```
Comment (jvieiro) [inline cardano-rosetta-server/src/server/utils/cardano/transactions-processor.ts:135]: This changes should be reverted

**Review by jvieiro — CHANGES_REQUESTED** [review]

File: cardano-rosetta-server/src/server/utils/cardano/operations-processor.ts [hunk 3]
```diff
@@ -182,6 +183,25 @@ const processOperationCertification = (
   };
 };
 
+const processPoolRetirement = (
+  logger: Logger,
+  operation: Components.Schemas.Operation
+): { certificate: CardanoWasm.Certificate; poolKeyHash: string } => {
+  logger.info(`[processPoolRetiring] About to process operation of type ${operation.type}`);
+  // eslint-disable-next-line camelcase
```
[the remark this answers is not in the document]
Comment (abarretoatix) [reply cardano-rosetta-server/src/server/utils/cardano/operations-processor.ts:191]: Not anymore, removed it

File: cardano-rosetta-server/src/server/utils/cardano/transactions-processor.ts [hunk 4]
```diff
@@ -295,14 +308,39 @@ const parseCertsToOperations = (
...
-        parsedOperations.push(parsedOperation);
+        parsedOperations.push({ ...parsedOperation, account: certOperation.account });
       }
     }
   }
 
   return parsedOperations;
 };
 
+const parsePoolRetirementToOperations = (
```
Comment (jvieiro) [inline cardano-rosetta-server/src/server/utils/cardano/transactions-processor.ts:319]: We can delete this function

File: cardano-rosetta-server/src/server/utils/cardano/transactions-processor.ts [hunk 5]
```diff
@@ -245,25 +244,39 @@ const parseCertToOperation = (
...
-
 const parseCertsToOperations = (
   logger: Logger,
   transactionBody: CardanoWasm.TransactionBody,
   certOps: Components.Schemas.Operation[],
-  certsCount: number,
   network: number
 ): Components.Schemas.Operation[] => {
   const parsedOperations = [];
+  const certsCount = certOps.length;
```
Comment (jvieiro) [inline cardano-rosetta-server/src/server/utils/cardano/transactions-processor.ts:279]:
```suggestion
  const certsCount = transactionBody.certs()?.len() || 0;
```

**Review by jvieiro — CHANGES_REQUESTED** [review]

**Pull request merged**